GH-533: Add ALP (Adaptive Lossless floating-Point) encoding specification (#557)
apache/parquet-format: Encodings.md

This file contains the specification of all supported encodings.

> | Encoding | ID | Supported Types |

Reviewer: Something seems strange to me about this chart. The version on main

https://github.com/apache/parquet-format/blob/96edf77704b60b6f3ca2232c218c64eff6c874d3/Encodings.md?plain=1#L28-L40

Doesn't appear in this document.

I would expect this PR that adds the new ALP encoding to add a new entry to the existing table, rather than add an entirely new table. 

Maybe we need to merge up from main to this branch (I am happy to do this and I think I have the permissions to push to your branch, but I don't want to mess things up on you)

Author: You're right, this branch was forked before the summary table was added in #550. Rebased onto upstream/master and added ALP as a row to the existing table.

| [Delta-length byte array](#DELTALENGTH)          | DELTA_LENGTH_BYTE_ARRAY = 6                               | BYTE_ARRAY                                        |
| [Delta Strings](#DELTASTRING)                    | DELTA_BYTE_ARRAY = 7                                      | BYTE_ARRAY, FIXED_LEN_BYTE_ARRAY                  |
| [Byte Stream Split](#BYTESTREAMSPLIT)            | BYTE_STREAM_SPLIT = 9                                     | INT32, INT64, FLOAT, DOUBLE, FIXED_LEN_BYTE_ARRAY |

> | [ALP](#ALP)                                      | ALP = 10                                                  | FLOAT, DOUBLE                                     |

Reviewer 2: What about FLBA and float16 datatype? Or should we even support fp8 or other types in the future? ( This can be not in this pr, just curious that)

Reviewer: It would be interesting to research as a next step in my opinion

The fixed length byte array (FLBA) seems unlikely to benefit from ALP (which is specifically designed to recognize values that are encoded using floating point that could be more efficiently represented as fixed precision decimals

FP16/FP8 could be helped, for decimal data that is stored in that type of column; However I think those types are often used for ML type usecases where the data uses the full range of the types, as I understand

Reviewer 3: I've raised the same question in the google doc earlier. FP16 and FP8 have fewer bits so the benefit may not be that significant compared to regular float and double. But I agree that we can discuss these in the future to not expand and thus block current spec design.

```
Bytes  AA 00 A3 BB 11 B4 CC 22 C5 DD 33 D6
```

> 

Reviewer 4: I think there are now links at the top of file summarizing applicability of encoding.

Author: Done — added an encoding applicability table at the top of the file with all encodings and their supported types.

<a name="ALP"></a>
### Adaptive Lossless floating-Point: (ALP = 10)

Supported Types: FLOAT, DOUBLE

This encoding is adapted from the paper
["ALP: Adaptive Lossless floating-Point Compression"](https://dl.acm.org/doi/10.1145/3626717)
by Afroozeh and Boncz (SIGMOD 2024).

ALP works by converting floating-point values to integers using decimal scaling,
then applying Frame of Reference (FOR) encoding and bit-packing. Values that
cannot be losslessly converted are stored as exceptions. The encoding achieves
high compression for decimal-like floating-point data (e.g., monetary values,

> sensor readings) while remaining fully lossless.

Reviewer: You have more summary at the end of the encoding, but I think a few more sentences in the intro would help people understand this more easily

```suggestion
ALP works by converting floating-point values to integers using decimal scaling,
then applying Frame of Reference (FOR) encoding and bit-packing. Values that
cannot be losslessly converted are stored as exceptions. The encoding achieves
high compression for decimal-like floating-point data (e.g., monetary values,
sensor readings) while remaining fully lossless. Values do not depend on
each other, which enables quick random access and parallel encode/decode. 
```

Author: Agreed — added sentence about value independence enabling random access and parallel encode/decode.

#### Overview

ALP encoding consists of a page-level header followed by an offset array and one
or more encoded vectors (batches of values). Each vector contains up to

> `vector_size` elements (default 1024).

Reviewer: I think it would help here to also define exponent (e) and factor (f) as they are referred to in the compression pipeline section as well as briefly mention the existence of exceptions here before they are described later in the spec

```suggestion
ALP encoding consists of a page-level header followed by an offset array and one
or more encoded vectors (batches of values). Each vector contains up to
`vector_size` elements (default 1024), transformed first using an `exponent` (`e`) 
and `factor` (`f`) and then encoded using frame-of-reference bit-packing.  Values
which can not be encoded losslessly, are stored as exceptions at the end of each vector. 
```

Author: Good call — introduced exponent, factor, and exceptions in the opening paragraph itself so the reader has the mental model before hitting the layout details.

Reviewer 3: I think moving this description to the top is slightly different with what @alamb has proposed. Here we can exactly explain how each vector is organized, however at the beginning readers have no idea what granularity is.

Author: Hi @wgtmac ,
Might have misunderstood you.
But are you suggesting we need to change the working here. Or we are good in the current state.

Reviewer 3: I think it is better to just apply @alamb's suggestion here so these concepts are not scattered and duplicated. Less is more.

Reviewer 3: > ALP encoding consists of a page-level header followed by an offset array and one or more encoded vectors (batches of values).

No strong opinion but I would like to clarify that ALP is same as other encodings that they apply to each data page. So here `page-level header` might be a little bit misleading (at least to me).

How about explicitly saying that `For each data page, ALP encoding consists of a header followed by an offset array and one or more encoded vectors (batches of values).`?

```
+-------------+-----------------------------+--------------------------------------+
|   Header    |        Offset Array         |            Vector Data               |
|  (7 bytes)  |   (num_vectors * 4 bytes)   |            (variable)                |
+-------------+------+------+-----+---------+----------+----------+-----+----------+
| Page Header | off0 | off1 | ... | off N-1 | Vector 0 | Vector 1 | ... | Vec N-1  |
|  (7 bytes)  | (4B) | (4B) |     |  (4B)   |(variable)|(variable)|     |(variable)|
+-------------+------+------+-----+---------+----------+----------+-----+----------+
```

The compression pipeline for each vector is:

```
                    Input: float/double array
                              |
                              v
    +----------------------------------------------------------+
    |  1. SAMPLING & PRESET GENERATION                         |

>     |     Sample vectors from column chunk                     |
>     |     Try all (exponent, factor) combinations              |
>     |     Select best k combinations for preset                |
>     +----------------------------------------------------------+

Reviewer: As I think was mentioned on the google doc spec, I think it would be better not to imply here that all implementations have to pick the exponent / factor by exhaustive search on a sample. It would be fine to suggest that as an implementation, but I think it is a particular algorithm that overly constrains the spec and could lead to future arguments if the spec requires this algorithm or not

Also technically I think the sample / presets are done one per input column in the paper, not once per vector (as stated above)

Also I think it would help to explicitly say exponent = e and factor = f here too.

Something like this perhaps

```suggestion
    +----------------------------------------------------------+
    |  1. Choose exponent (e) and factor (f).                  |
    +----------------------------------------------------------+
```

Author: Agreed — simplified the pipeline to 'Choose parameters'. The sampling/preset strategy remains in the Parameter Selection section as a SHOULD recommendation.

    |  2. DECIMAL ENCODING                                     |
    |     encoded[i] = round(value[i] * 10^e * 10^(-f))       |
    |     Detect exceptions where decode(encode(v)) != v       |
    +----------------------------------------------------------+
                              |
                              v
    +----------------------------------------------------------+
    |  3. FRAME OF REFERENCE (FOR)                             |
    |     min_val = min(encoded[])                             |
    |     delta[i] = encoded[i] - min_val                      |
    +----------------------------------------------------------+
                              |
                              v
    +----------------------------------------------------------+
    |  4. BIT PACKING                                          |
    |     bit_width = ceil(log2(max_delta + 1))                |
    |     Pack each delta into bit_width bits                  |
    +----------------------------------------------------------+
                              |
                              v
                   Output: Serialized vector bytes
```

#### Page Layout

##### Header (7 bytes)

> All multi-byte values are little-endian.

Reviewer 4: ```suggestion
All multi-byte values are stored in little-endian order.
```

Author: Done

```
 Byte:    0              1               2              3    4    5    6
       +----------------+---------------+--------------+----+----+----+----+
       | compression    | integer       | log_vector   |     num_elements  |
       | _mode          | _encoding     | _size        |     (int32 LE)    |
       +----------------+---------------+--------------+----+----+----+----+
```

| Offset | Field | Size | Type | Description |
|--------|-------|------|------|-------------|

> | 0 | compression_mode | 1 byte | uint8 | Compression mode (must be 0 = ALP) |

Reviewer 4: we think ALP-RD fits in well here?  I forget what the extension point is and why we were OK keeping this field, but not version.

Author: changing the code to read ALP's newer laid out bits would be difficult to apply across all implementations. but AlpRD is a completely newer implementations.

Added a note to the field description.

| 1 | integer_encoding | 1 byte | uint8 | Integer encoding (must be 0 = FOR + bit-packing) |

> | 2 | log_vector_size | 1 byte | uint8 | log2(vector\_size). Must be in \[3, 15\]. Default: 10 (vector size 1024) |

Reviewer 4: ```suggestion
| 2 | log_vector_size | 1 byte | uint8 | log2(vector\_size). Must be in \[3, 15\]. Recommended default: 10 (vector size 1024) |
```

Author: done

> | 2 | log_vector_size | 1 byte | uint8 | log2(vector\_size). Must be in \[3, 15\]. Default: 10 (vector size 1024) |

Reviewer 4: ```suggestion
| 2 | log_vector_size | 1 byte | uint8 | log2(vector\_size). Must be in the inclusive range: \[3, 15\]. Default: 10 (vector size 1024) |
```

Author: done

| 3 | num_elements | 4 bytes | int32 | Total number of floating-point values in the page |

The number of vectors is `ceil(num_elements / vector_size)`. The last vector may
contain fewer than `vector_size` elements.

**Note:** The number of elements per vector and the packed data size are NOT stored
in the header. They are derived:

> * Elements per vector: `vector_size` for all vectors except the last, which may be smaller.

Reviewer 4: This seems like a little bit of a strange callout since it is covered on line 457 explicitly.  and log_vector_size is stored in the header?

Reviewer: I agree this is redundant -- I think leaving the note is valuable clarification

> **Note:** The number of elements per vector and the packed data size are NOT stored
in the header, they are derived

However the other items are not

Author: Removed the packed data size bullet, kept the elements-per-vector clarification.

> * Packed data size: `ceil(num_elements_in_vector * bit_width / 8)`.

Reviewer 4: bit_width isn't in the header either, so it is a little strange to call ths out here?

Reviewer: I also hihglighted this as confusing when reading the spec. I recommend removing this sentence as the packed data size is covered in the "Vector Format" section below

Author: comment regarding packed data size has been removed.

##### Offset Array

Immediately following the header is an array of `num_vectors` little-endian uint32
values. Each offset gives the byte position of the corresponding vector's data,
measured from the start of the offset array itself.

> The first offset equals `num_vectors * 4` (pointing just past the offset array).

Reviewer 4: ```suggestion
The first offset always equals `num_vectors * 4` (pointing just past the offset array).
```
Lets be explicitly here that we don't support padding.

Author: done

Each subsequent offset equals the previous offset plus the stored size of the
previous vector.

##### Vector Format

Each vector is self-describing and contains the encoding parameters, FOR metadata,
bit-packed encoded values, and exception data.

```
+-------------------+-----------------+-------------------+---------------------+-------------------+
|      AlpInfo      |     ForInfo     |   PackedValues    | ExceptionPositions  | ExceptionValues   |
|     (4 bytes)     | (5B or 9B)      |    (variable)     |     (variable)      |    (variable)     |
+-------------------+-----------------+-------------------+---------------------+-------------------+
```

Vector header sizes:
| Type   | AlpInfo | ForInfo | Total Header |
|--------|---------|---------|--------------|
| FLOAT  | 4 bytes | 5 bytes | 9 bytes      |
| DOUBLE | 4 bytes | 9 bytes | 13 bytes     |

Data section sizes:
| Section             | Size Formula                | Description                  |
|---------------------|-----------------------------|------------------------------|

> | PackedValues        | ceil(N * bit\_width / 8)    | Bit-packed delta values      |

Reviewer 4: ```suggestion
| PackedValues        | ceil(`vector_size` * bit\_width / 8)    | Bit-packed delta values      |
```

Author: done

| ExceptionPositions  | num\_exceptions * 2 bytes   | uint16 indices of exceptions |

> | ExceptionValues     | num\_exceptions * sizeof(T) | Original float/double values |

Reviewer 4: ```suggestion
| ExceptionValues     | num\_exceptions * sizeof(encoded type) (float=4 and double=8) | Original float/double values |
```

Author: done

       | frame_of_reference | bit_width |
       |    (int32 LE)      |  (uint8)  |
       +----+----+----+----+-----------+
```

| Offset | Field | Size | Type | Description |
|--------|-------|------|------|-------------|
| 0 | frame_of_reference | 4 bytes | int32 | Minimum encoded integer in the vector |
| 4 | bit_width | 1 byte | uint8 | Bits per packed value. Range: \[0, 32\]. |

###### ForInfo for DOUBLE (9 bytes)

```
 Byte:    0    1    2    3    4    5    6    7       8
       +----+----+----+----+----+----+----+----+-----------+
       |          frame_of_reference           | bit_width |
       |              (int64 LE)               |  (uint8)  |
       +----+----+----+----+----+----+----+----+-----------+
```

| Offset | Field | Size | Type | Description |
|--------|-------|------|------|-------------|
| 0 | frame_of_reference | 8 bytes | int64 | Minimum encoded long in the vector |
| 8 | bit_width | 1 byte | uint8 | Bits per packed value. Range: \[0, 64\]. |

###### PackedValues

The FOR-encoded deltas, bit-packed into `ceil(num_elements_in_vector * bit_width / 8)` bytes.
Values are packed from the least significant bit of each byte to the most significant bit,

> in groups of 8 values, using the same bit-packing order as the

Reviewer 4: Where does the group of 8 values come in?  Wouldn't this messup the number of bytes math?

Author: Good catch — removed 'groups of 8' phrasing. Now simply references the same LSB-first packing order as RLE/Bit-Packing Hybrid.

(I think RleBitPackHybrid was in my mind at that point :) )

[RLE/Bit-Packing Hybrid](#RLE) encoding.

If `bit_width` is 0, no bytes are stored (all deltas are zero, meaning all encoded
integers are equal to `frame_of_reference`).

###### ExceptionPositions

An array of `num_exceptions` little-endian uint16 values, each giving
the 0-based index within the vector of an exception value.

###### ExceptionValues

An array of `num_exceptions` values in the original floating-point type
(4 bytes little-endian IEEE 754 for FLOAT, 8 bytes for DOUBLE), stored in
the same order as the corresponding positions.

#### Encoding

##### Encoding Formula

```
+-------------------------------------------------------------------+
|                                                                   |

> |   encoded = round( value  *  10^e  *  10^(-f) )                  |

Reviewer 4: ```suggestion
|   encoded = fast_round( value  *  10^e  *  10^(-f) )                  |
```

Author: done

|                                                                   |
|   decoded = encoded  *  10^f  *  10^(-e)                          |
|                                                                   |
+-------------------------------------------------------------------+
```

The encoding uses two separate multiplications (not a single multiplication by
`10^(e-f)`, and not division) to ensure that implementations produce identical
floating-point rounding across languages. The powers of 10 MUST be stored as

> precomputed floating-point constants (i.e., literal values like `1e-3f`), not

Reviewer 4: Why can't they be precomputed at runtime?

Author: You're right — this shouldn't mandate literals vs runtime computation. Reworded to require that encoder and decoder use identical power-of-10 values, without prescribing how they're obtained.

computed at runtime.

##### Fast Rounding

The rounding function uses a "magic number" technique for branchless rounding:

| Type   | Magic Number                      | Formula                          |
|--------|-----------------------------------|----------------------------------|

> | FLOAT  | 2^22 + 2^23 = 12,582,912         | `(int)((value + magic) - magic)` |

Reviewer 4: ```suggestion
| FLOAT  | 2^22 + 2^23 = 12,582,912         | `(int32_t)((value + magic) - magic)` |
```

Author: done

> | DOUBLE | 2^51 + 2^52 = 6,755,399,441,055,744 | `(long)((value + magic) - magic)` |

Reviewer 4: ```suggestion
| DOUBLE | 2^51 + 2^52 = 6,755,399,441,055,744 | `(int64_t)((value + magic) - magic)` |
```

Author: done

```
Bytes  AA 00 A3 BB 11 B4 CC 22 C5 DD 33 D6
```

<a name="ALP"></a>
### Adaptive Lossless floating-Point: (ALP = 10)

Supported Types: FLOAT, DOUBLE

This encoding is adapted from the paper
["ALP: Adaptive Lossless floating-Point Compression"](https://dl.acm.org/doi/10.1145/3626717)
by Afroozeh and Boncz (SIGMOD 2024).

ALP works by converting floating-point values to integers using decimal scaling
(controlled by an *exponent* `e` and *factor* `f`), then applying Frame of
Reference (FOR) encoding and bit-packing. Values that cannot be losslessly
converted are stored separately as *exceptions*. The encoding achieves high
compression for decimal-like floating-point data (e.g., monetary values, sensor
readings) while remaining fully lossless. Each value is encoded independently,

> enabling random access to individual vectors and parallel encode/decode.

Reviewer: ```suggestion
enabling random access to individual values and parallel encode/decode.
```

Author: done

> #### Overview

Reviewer 5: The algorithm description is so long that I think it should be moved to a separate file that we would link to here. In this file we would just keeping the description that is before the Overview.

Author: Yes this makes sense. I too thought it became long and having a separate file would be good.
Let me take a stab at it after I address the above comments and get approval. Else the comment threads on github will get lost.

Reviewer: Maybe similar to how it is done for bloom filter: https://github.com/apache/parquet-format/blob/master/BloomFilter.md

Reviewer: I also think it would be ok to if we moved the content to a separate file as a follow on PR (after the spec change is approved) as in my mind moving the content to a separate file does not affect the content of the spec

Reviewer: It seems we should do this as a follow on PR. I can do so after this PR is merged

Author: Got it. Let's address it in a follow up PR

Reviewer: Filed a ticket to track
- https://github.com/apache/parquet-format/issues/602

Reviewer: PR to move the content:
- https://github.com/apache/parquet-format/pull/608

integers are equal to `frame_of_reference`).

###### ExceptionPositions

An array of `num_exceptions` little-endian uint16 values, each giving
the 0-based index within the vector of an exception value.

###### ExceptionValues

An array of `num_exceptions` values in the original floating-point type
(4 bytes little-endian IEEE 754 for FLOAT, 8 bytes for DOUBLE), stored in
the same order as the corresponding positions.

#### Encoding

##### Encoding Formula

```
+-------------------------------------------------------------------+
|                                                                   |
|   encoded = fast_round( value  *  10^e  *  10^(-f) )             |
|                                                                   |
|   decoded = encoded  *  10^f  *  10^(-e)                          |
|                                                                   |
+-------------------------------------------------------------------+
```

The encoding uses two separate multiplications (not a single multiplication by
`10^(e-f)`, and not division) to ensure that implementations produce identical
floating-point rounding across languages. Implementations must ensure that the

> encoder and decoder use identical power-of-10 values for a given exponent.

Reviewer: I don't understand what "use identical power-of-10 values" means. **ALL** encoders and decoders must  to use the exact same values (and floating point arithmetic) as I understand it.

Author: Yes, what you said. Reworded to: "All implementations MUST use the exact same floating-point arithmetic and power-of-10 constants to guarantee cross-language interoperability."

##### Fast Rounding

> The rounding function uses a "magic number" technique for branchless rounding:

Reviewer: I think it would help to use consistent terminology here to above. 

```suggestion
The `fast_round` function uses a "magic number" technique for branchless rounding.

`fast_round(value)` is defined as follows:
```

Author: Done. Updated to use fast_round consistently and added "is defined as follows".

| Type   | Magic Number                      | Formula                          |
|--------|-----------------------------------|----------------------------------|
| FLOAT  | 2^22 + 2^23 = 12,582,912         | `(int32_t)((value + magic) - magic)` |
| DOUBLE | 2^51 + 2^52 = 6,755,399,441,055,744 | `(int64_t)((value + magic) - magic)` |

> For negative values, the signs are reversed: `(int32_t)((value - magic) + magic)` for FLOAT, `(int64_t)((value - magic) + magic)` for DOUBLE.

Reviewer: I don't understand this sentence. The C++ implementation does not seem to change sign order from what I can tell:

https://github.com/apache/arrow/pull/48345/changes#diff-f9ab708cab94060b4067fff0a6739e9c3751b450422115663b2bd0badfcc748bR487-R490

```c++
 /// \brief Convert a float to an int without rounding
  static inline auto FastRound(T n) -> SignedExactType {
    n = n + Constants::kMagicNumber - Constants::kMagicNumber;
    return static_cast<SignedExactType>(n);
  }
```

Also the ["Fast Rounding"](https://ir.cwi.nl/pub/33334/33334.pdf) section of the ALP paper doesn't mention sign reversal that I can find

Author: You're right -- the sign reversal was incorrect. The C++ implementation uses a single formula for both positive and negative values (n + magic - magic). Removed the paragraph.

##### Parameter Selection

The encoder selects the (exponent, factor) pair that minimizes exceptions.
Valid combinations satisfy 0 &le; factor &le; exponent:

| Type   | Max Exponent | Total Combinations |
|--------|--------------|--------------------|
| FLOAT  | 10           | 66                 |
| DOUBLE | 18           | 190                |

To avoid the cost of exhaustive search on every vector, implementations
SHOULD use sampling to select up to 5 candidate (exponent, factor)
combinations (the "encoding preset") at the start of each column chunk.

> Each vector then searches only those 5 candidates.

Reviewer: I think *SHOULD* is too strong a word here. Maybe we could soften it to a suggestion

```suggestion
To avoid the cost of exhaustive search on every vector, implementations 
can use a sampling approach. One such approach, described in the paper,  is to
select up to 5 candidate (exponent, factor) combinations (the "encoding preset") 
at the start of each column chunk, and when encoding each vector,  
test each of the 5 candidates for the fewest exceptions. 
```

Author: Done

| Sample Size          | 256   | Values sampled per vector           |
| Max Combinations     | 5     | Best (e,f) pairs kept in preset     |
| Sample Vectors       | 8     | Vectors sampled per row group       |

##### Exception Detection

A value becomes an exception if any of the following is true:

| Condition          | Example                    | Reason                           |
|--------------------|----------------------------|----------------------------------|
| NaN                | `NaN`                      | Cannot convert to integer        |
| Infinity           | `+Inf`, `-Inf`             | Cannot convert to integer        |
| Negative zero      | `-0.0`                     | Would become `+0.0` after encoding |
| Out of range       | value * 10^e > INT32\_MAX  | Exceeds target integer limits    |
| Round-trip failure  | `0.333...` with e=1, f=0  | `decode(encode(v)) != v`         |

Exception values at positions in the vector are replaced with a placeholder
(the encoded integer of the first non-exception value, or 0 if all values
are exceptions) before FOR encoding. This keeps the FOR range tight.

> ##### Frame of Reference and Bit-Packing

Reviewer: This section has a worked example of FOR / bitpacking so perhaps we could write it like

```suggestion
##### Example: Frame of Reference and Bit-Packing
```

Author: Updated

> After decimal encoding and exception substitution:

Reviewer: ```suggestion
Given the following data after decimal encoding and exception substitution:
```

Author: Updated

                              |
                              v
    +----------------------------------------------------------+
    |  3. DECIMAL DECODING                                     |
    |     value[i] = encoded[i] * 10^factor * 10^(-exponent)   |
    +----------------------------------------------------------+
                              |
                              v
    +----------------------------------------------------------+
    |  4. PATCH EXCEPTIONS                                     |
    |     value[pos[j]] = exception_values[j]                  |
    +----------------------------------------------------------+
                              |
                              v
                  Output: Original float/double array
```

For each vector:

1. Read AlpInfo and ForInfo from the vector header.
2. Unpack `bit_width`-bit integers from PackedValues.
3. Add `frame_of_reference` to each unpacked integer.
4. Decode: multiply each integer by `10^factor` then by `10^(-exponent)`.
5. Patch exceptions: for each (position, value) in the exception arrays,
   overwrite the decoded output at that position with the stored value.

> #### Example 1: Simple Decimal Values

Reviewer: I reviewed this section and I think the content in the Examples is really nice but too verbose to include all of them in the spec. 

Specifically 
1. I think one fully worked example (with exceptions) would help understand how this works
2. The size analysis and comparison to PLAIN encoding or other encodings is too much detail for this spec (it makes the spec change significantly longer, and does not help someone implement the encoding, but rather decide when to use ALP / why it was added to the spec)
3. None of the examples shows a `f` factor other than zero -- it would be good to add an example with a non zero f

Thus I propose:
1. Reduce to a single example (I think Example 2)
2. Update the example with `f != 0`
3. Move the other examples and comparisons to the blog post we will write for ALP (simplarly to https://parquet.apache.org/blog/2026/02/27/variant-type-in-apache-parquet-for-semi-structured-data/ or https://parquet.apache.org/blog/2026/02/13/native-geospatial-types-in-apache-parquet/)

Would it be ok if I started writing a blog post about this feature (and thus we had a place to move these examples)?

Reviewer: Here is a ticket that describes the ALP blog:
- https://github.com/apache/parquet-site/issues/175

Author: Done. Consolidated into one worked example using double with (e=4, f=3) and a NaN exception. Removed Examples 1 and 3.

| **Total**           |                                        | **23 bytes** |

#### Example 3: Monetary Data (1024 values)

1024 price values ranging from $0.01 to $999.99 (e.g., product prices).

Optimal encoding: (exponent=2, factor=0)

| Metric        | Value       | Calculation                          |
|---------------|-------------|--------------------------------------|
| Exponent      | 2           | Multiply by 100 for 2 decimal places |
| Factor        | 0           | No additional scaling needed         |
| Encoded range | 1 to 99,999 | $0.01 -> 1, $999.99 -> 99999         |
| FOR min       | 1           | Assuming $0.01 is present            |
| Delta range   | 0 to 99,998 | After FOR subtraction                |
| Bit width     | 17          | ceil(log2(99999)) = 17 bits          |
| Packed size   | 2,176 bytes | ceil(1024 * 17 / 8)                  |

**Size Comparison:**

| Encoding      | Size         | Ratio               |
|---------------|--------------|----------------------|
| PLAIN (float) | 4,096 bytes  | 1.0x                 |
| ALP           | ~2,185 bytes | 0.53x (47% smaller)  |

> #### Characteristics

Reviewer: as above, this is great stuff but probably doesn't belong in the spec

Author: removed

* Scientific measurements with limited decimal places
* GPS coordinates and geographic data
* Normalized scores and percentages

**Worst case scenarios:**

* Random floating-point values (high exception rate)
* High-precision scientific data (many decimal places)
* Data with many special values (NaN, Inf)
* Very small datasets (header overhead dominates)

**Comparison with other encodings:**

| Encoding            | Type Support | Compression | Best For            |
|---------------------|--------------|-------------|---------------------|
| PLAIN               | All          | None        | General purpose     |
| BYTE\_STREAM\_SPLIT | Float/Double | Moderate    | Random floats       |
| ALP                 | Float/Double | High        | Decimal-like floats |
| DELTA\_BINARY\_PACKED | Int32/Int64 | High       | Sequential integers |

Unlike [Byte Stream Split](#BYTE_STREAM_SPLIT), ALP does not require a subsequent
compression step to achieve size reduction -- the bit-packing directly reduces the
encoded size. However, ALP and Byte Stream Split can be complementary: ALP
exploits decimal structure while Byte Stream Split exploits byte-level correlation.

> #### Size Calculations

Reviewer: I don't think we need to add add to the Parquet spec -- it isn't required to implement the encoder/decoder for ALP.

Author: removed

##### Vector Size Formula

```
vector_bytes = vector_header_size                  // FLOAT: 9, DOUBLE: 13
             + ceil(num_elements * bit_width / 8)  // packed values
             + num_exceptions * 2                  // exception positions (uint16)
             + num_exceptions * sizeof(T)          // exception values (4 or 8)
```

##### Page Size Formula

```
page_bytes = 7                                   // page header
           + num_vectors * 4                     // offset array
           + sum(vector_bytes for each vector)   // all vectors
```

> #### Constants Reference

Reviewer: This appears to be C/++ implementation details -- I am not sure it adds any value to the spec

Author: removed

integers are equal to `frame_of_reference`).

###### ExceptionPositions

An array of `num_exceptions` little-endian uint16 values, each giving
the 0-based index within the vector of an exception value.

###### ExceptionValues

An array of `num_exceptions` values in the original floating-point type
(4 bytes little-endian IEEE 754 for FLOAT, 8 bytes for DOUBLE), stored in
the same order as the corresponding positions.

#### Encoding

##### Encoding Formula

```
+-------------------------------------------------------------------+
|                                                                   |
|   encoded = fast_round( value  *  10^e  *  10^(-f) )             |
|                                                                   |
|   decoded = encoded  *  10^f  *  10^(-e)                          |
|                                                                   |
+-------------------------------------------------------------------+
```

The encoding uses two separate multiplications (not a single multiplication by
`10^(e-f)`, and not division) to ensure that implementations produce identical
floating-point results. All implementations MUST use the exact same floating-point

> arithmetic and power-of-10 constants to guarantee cross-language interoperability.

Reviewer 6: This requirement is not actionable without specifying the actual constant values. Powers of 10 like `10^(-3)` are not exactly representable in IEEE 754, and different methods of computing them (e.g., `1.0/1000.0` vs the compile-time literal `1e-3` vs `pow(10, -3)`) can produce different bit patterns.

Suggestion: either (a) provide a table of required constants with their exact IEEE 754 hex representations, or (b) specify that constants MUST match the values produced by the standard decimal-to-binary conversion of the literals `1e0`, `1e1`, ..., `1e18` and `1e-0`, `1e-1`, ..., `1e-18` as defined by IEEE 754-2008 §5.12.2. This makes the requirement unambiguous and testable across languages.

Author: Thanks for the review. The concern about `1.0/1000.0` vs `1e-3` producing different bit patterns is not actually valid — both operands are exactly   representable in IEEE 754, and division is a required correctly-rounded operation, so the result is bit-identical to the literal. The real risk you're identifying is `pow()`, which is not required to be correctly rounded (it's only a "recommended" operation even in IEEE 754-2019).
                                                                                                                                                           
  That said, I agree the spec should be explicit. I've added a clarification that the constants MUST match the correctly-rounded IEEE 754 decimal-to-binary conversion of the literals `1e0`..`1e18` and `1e-1`..`1e-18` per §5.12.2, and that implementations MUST NOT compute these via `pow()`. This makes the requirement unambiguous and testable without needing a hex table — any conforming implementation will produce identical bit patterns from these short  decimal literals.

Reviewer 7: Do we have a test case or two where we know `pow()` produces an incorrect value in some cases? Then we could add a table of test values to help catch the issue.

Reviewer: > This requirement is not actionable without specifying the actual constant values. 

I don't agree that this requirement is not actionable (we have at least two implementations that didn't seem to have any trouble figuring out `1e0` and `1e2` man. 

> (b) specify that constants MUST match the values produced by the standard decimal-to-binary conversion of the literals 1e0, 1e1, ..., 1e18 and 1e-0, 1e-1, ..., 1e-18 as defined by

I think that would be ok, but to be honest I don't think any more clarification is needed here

Reviewer: > Do we have a test case or two where we know pow() produces an incorrect value in some cases? Then we could add a table of test values to help catch the issue.

Given the spec doesn't call for `pow()` I the the discussion about the exact numeric representation of `pow()` is somewhat distracting -- I recommend we leave the spec as is

##### Fast Rounding

The `fast_round` function uses a "magic number" technique for branchless rounding.

`fast_round(value)` is defined as follows:

| Type   | Magic Number                      | Formula                          |
|--------|-----------------------------------|----------------------------------|
| FLOAT  | 2^22 + 2^23 = 12,582,912         | `(int32_t)((value + magic) - magic)` |

> | DOUBLE | 2^51 + 2^52 = 6,755,399,441,055,744 | `(int64_t)((value + magic) - magic)` |

Reviewer 6: The formula shown here only covers non-negative values. For negative values, the expression `value + magic` may fall below the binade `[2^52, 2^53)` (for double), entering a region where ULP = 0.5 instead of 1.0, which produces incorrect rounding.

Known implementations (C++/DuckDB, Java) use sign branching:
```
if value >= 0:
    result = (int64_t)((value + magic) - magic)
else:
    result = (int64_t)((value - magic) + magic)
```

While the round-trip exception check prevents data corruption (incorrectly rounded values become exceptions), omitting sign branching causes a higher exception rate for datasets with negative values, degrading compression ratios.

Suggestion: document both branches in the spec table, or at minimum add a note that implementations SHOULD use sign branching for negative values.

Author: Good catch — you're right. I've updated the spec to document both branches:
                                                                                                                                                           
      if value >= 0:  (intN_t)((value + magic) - magic)                                                                                                    
      if value <  0:  (intN_t)((value - magic) + magic)                                                                                                    
                                                                                                                                                           
I also added a paragraph explaining why: the technique relies on landing in a binade where ULP = 1.0, which only works for non-negative values with  +magic and negative values with -magic. Without sign branching, negative values with magnitude beyond 2^22 (float) or 2^51 (double) fall into a lower binade and round incorrectly — silently increasing the exception rate.

Reviewer: > Known implementations (C++/DuckDB, Java) use sign branching:

@iemejia can you please provide  example / evidence for this this statement?

The three implementatons I could find all use brancless code (I found not find any check for sign) -- they simply add then subtract the magic number

# DuckDB's implementation of ALP (C++)
https://github.com/duckdb/duckdb/blob/68d73b7c3a85a8907909cf6d4b6533f83a45496d/src/include/duckdb/storage/compression/alp/algorithm/alp.hpp#L118-L134

# SpiralDB's implementation of ALP (Rust) 
https://github.com/spiraldb/alp/blob/d797ba56f980f83657bf6b812f88714ebe3ed674/src/alp/mod.rs#L70-L74

# cwida/ALP (reference implementation from the paper authors) 
https://github.com/cwida/ALP/blob/31ca0ed11c93c99d3f5b5c30e01a3e1c3832d3ce/include/alp/encoder.hpp#L87

Reviewer: > Without sign branching, negative values with magnitude beyond 2^22 (float) or 2^51 (double) fall into a lower binade and round incorrectly — silently increasing the exception rate.

Can you please provide some examples where encoding using the signed version would correctly recover the intended value but the unsigned version would not?

What is the evidence that the exception rate would be higher?

Reviewer: I made a plaground to show how to test various combinations / how fast_round works: https://github.com/alamb/fast_round\

(update: I can't find any reason why we would want to add the sign test)

##### Parameter Selection

> The encoder selects the (exponent, factor) pair that minimizes exceptions.

Reviewer 6: In practice, minimizing exception count alone is not optimal. A parameter pair with slightly more exceptions but a much smaller bit-width (tighter FOR range) can produce smaller output. The actual optimization target should be estimated encoded size, accounting for both bit-width and exception overhead.

Since this is an encoder-only concern (decoders are agnostic to the selection strategy), the spec should clarify that any valid (e, f) pair produces a correct encoding — the choice only affects compression ratio. Suggested rewording:

> The encoder SHOULD select the (exponent, factor) pair that produces the smallest encoded output. A simple heuristic is to minimize exception count; a more precise approach accounts for both bit-width and exception overhead.

Author: Good point — accepted. I've reworked the section to:      
                                                                                                                                                           
  1. Lead with the correctness statement: any valid (e,f) pair produces a correct encoding, the decoder is agnostic, and the exception mechanism guarantees round-trip fidelity regardless.                                                                                                                         
  2. Use SHOULD for the optimization guidance and frame "minimize exceptions" as one heuristic rather than the goal. The actual target is smallest encoded size, accounting for both bit-width and exception overhead.                                                                                              
  3. Updated the sampling paragraph accordingly.

select up to 5 candidate (exponent, factor) combinations (the "encoding preset")
at the start of each column chunk, and when encoding each vector,
test each of the 5 candidates for the fewest exceptions.

Suggested sampling parameters (from the paper):

| Parameter            | Value | Description                         |
|----------------------|-------|-------------------------------------|
| Sample Size          | 256   | Values sampled per vector           |
| Max Combinations     | 5     | Best (e,f) pairs kept in preset     |
| Sample Vectors       | 8     | Vectors sampled per row group       |

##### Exception Detection

A value becomes an exception if any of the following is true:

| Condition          | Example                    | Reason                           |
|--------------------|----------------------------|----------------------------------|
| NaN                | `NaN`                      | Cannot convert to integer        |
| Infinity           | `+Inf`, `-Inf`             | Cannot convert to integer        |
| Negative zero      | `-0.0`                     | Would become `+0.0` after encoding |

> | Out of range       | value * 10^e > INT32\_MAX  | Exceeds target integer limits    |

Reviewer 6: This condition is under-specified in two ways:

1. For DOUBLE, the target integer type is int64, not int32. The condition should reference INT64_MAX for doubles.
2. The actual safe casting limit is not exactly `INT32_MAX` / `INT64_MAX` but the largest floating-point value that can be safely converted to the target integer type without undefined behavior. For float→int32 this is approximately `2147483520.0f` (not `2147483647`), and for double→int64 approximately `9223372036854774784.0` (not `9223372036854775807`). These differ because not all integers near the max are exactly representable in the source float/double type.

Suggestion: either specify the exact limits for each type, or reword to: "the scaled value exceeds the range that can be losslessly represented in the target integer type (int32 for FLOAT, int64 for DOUBLE)."

Author: Good catch on both points. Fixed — the table now reads:   
                                                                                                                                                           
  | Out of range | scaled value outside int32 (FLOAT) or int64 (DOUBLE) | Exceeds target integer type range                                               
                                                                                                                                                           
  This avoids both the int32-only bug and the subtlety around exact casting limits (e.g., 2147483520.0 being the actual safe boundary for float→int32). The round-trip check remains the authoritative correctness condition regardless.

For each vector:

1. Read AlpInfo and ForInfo from the vector header.
2. Unpack `bit_width`-bit integers from PackedValues.
3. Add `frame_of_reference` to each unpacked integer.
4. Decode: multiply each integer by `10^factor` then by `10^(-exponent)`.
5. Patch exceptions: for each (position, value) in the exception arrays,
   overwrite the decoded output at that position with the stored value.

#### Worked Example: Exceptions and Non-Zero Factor

**Input:** `double values[4] = { 1500.0, NaN, 2500.0, 333.3 }`

Best encoding found: (exponent=4, factor=3). This means:
`encoded = fast_round(value * 10^4 * 10^(-3)) = fast_round(value * 10)`

**Step 1: Decimal Encoding**

| Index | Value   | value * 10^4 * 10^(-3) | Rounded | Decoded: rounded * 10^3 * 10^(-4) | Exception? |
|-------|---------|------------------------|---------|------------------------------------|------------|
| 0     | 1500.0  | 15000.0                | 15000   | 1500.0                             | No         |
| 1     | NaN     | -                      | -       | -                                  | Yes (NaN)  |
| 2     | 2500.0  | 25000.0                | 25000   | 2500.0                             | No         |

> | 3     | 333.3   | 3333.0                 | 3333    | 333.3                              | No         |

Reviewer 6: The value `333.3` with `(exponent=4, factor=3)` may not survive a round-trip under strict IEEE 754 arithmetic:

- Encode: `333.3 * 10^4 * 10^(-3)` = `333.3 * 10000.0 * 0.001`. Since `0.001` is not exactly representable in IEEE 754 double, the product is approximately `3333.000000000000069...`, which rounds to `3333`.
- Decode: `3333 * 10^3 * 10^(-4)` = `3333 * 1000 * 0.0001`. Since `0.0001` is not exactly representable, the product may not bit-equal `333.3`.

If `decode(encode(333.3)) != 333.3` at the bit level, this value should be classified as an exception. The example may be using idealized math rather than actual IEEE 754 double semantics.

Suggestion: verify this specific round-trip in an actual IEEE 754 implementation, or replace `333.3` with a value that provably round-trips (e.g., `3000.0`, `500.0`, or any value where the scaling produces an exact result).

Author: Umm,

I verified this computationally — 333.3 with (e=4, f=3) round-trips exactly under IEEE 754 double arithmetic:                                            
                                                                                                                                                           
    Encode: 333.3 * 10000.0 = 3333000.0 (exact), * 0.001 = 3333.0 (exact)                                                                                  
    Decode: 3333 * 1000.0 = 3333000.0 (exact), * 0.0001 = 333.3 (bit-identical to original)                                                                
                                                                                                                                                           
 The intermediate value is exactly 3333.0, not ~3333.000000000000069 as suggested. This is precisely the property that the two-separate-multiplications  design guarantees: for values with d decimal digits where d ≤ (e - f), the round-trip is exact. Here 333.3 has d=1 digit and e-f=1, so the invariant   holds.                                                                                                                                                   
                                                                                                                                                           
  The concern that "0.001 is not exactly representable therefore the product is I think not right" doesn't follow — IEEE 754 correctly-rounded multiplication of  3333000.0 by the nearest double to 0.001 produces exactly 3333.0 in this case.

Reviewer 4: I think the concern might be 333.3 as text is not actually the value that gets stored as a double.  For instance if you look at (https://www.h-schmidt.net/FloatConverter/IEEE754.html) `333.29998779296875`.  I think the math still ends up working out here though.  Generally, ".3" is the canonical example of something that doesn't store exactly into floating point values.  I think the recommendation here would be use something like 333.25 which doesn't suffer from this problem.

Author: Ahh got it.

Changed the worked-example value from 333.3 to 333.5. 333.5 is exactly representable in IEEE 754 (667/2), so it sidesteps the ".3 isn't stored exactly" issue entirely while still using a non-zero factor.

Reference (FOR) encoding and bit-packing. Values that cannot be losslessly
converted are stored separately as *exceptions*. The encoding achieves high
compression for decimal-like floating-point data (e.g., monetary values, sensor
readings) while remaining fully lossless. Each value is encoded independently,
enabling random access to individual values and parallel encode/decode.

#### Overview

ALP encoding consists of a page-level header followed by an offset array and one
or more encoded vectors (batches of values). Each vector contains up to
`vector_size` elements (default 1024).

```
+-------------+-----------------------------+--------------------------------------+
|   Header    |        Offset Array         |            Vector Data               |
|  (7 bytes)  |   (num_vectors * 4 bytes)   |            (variable)                |
+-------------+------+------+-----+---------+----------+----------+-----+----------+
| Page Header | off0 | off1 | ... | off N-1 | Vector 0 | Vector 1 | ... | Vec N-1  |
|  (7 bytes)  | (4B) | (4B) |     |  (4B)   |(variable)|(variable)|     |(variable)|
+-------------+------+------+-----+---------+----------+----------+-----+----------+
```

> The compression pipeline for each vector is:

Reviewer 3: From the perspective of spec, do we really need to describe how encoder works here? IMHO, describing how to decode a page is sufficient. If we will split a dedicated file for ALP later, then it is absolutely a good idea to guide how to implement the encoder.

Reviewer 4: I think this will be restructured, but I think giving implementers a sketch of one way of encoding is a good idea (it should be clearly marked as one option).

Author: Pipeline now labeled informative, not normative; only the byte layout + decode path are normative.

```
                    Input: float/double array
                              |
                              v
    +----------------------------------------------------------+
    |  1. CHOOSE PARAMETERS                                    |
    |     Select (exponent, factor) pair for this vector       |
    +----------------------------------------------------------+
                              |
                              v
    +----------------------------------------------------------+
    |  2. DECIMAL ENCODING                                     |

>     |     encoded[i] = fast_round(value[i] * 10^e * 10^(-f))  |

Reviewer 4: haven't looked at the C++ change, but noticed something about branching there related to rounding, does this need to be updated?

Author: Added pointer to the Fast Rounding section after the pipeline

    |     Detect exceptions where decode(encode(v)) != v       |
    +----------------------------------------------------------+
                              |
                              v
    +----------------------------------------------------------+
    |  3. FRAME OF REFERENCE (FOR)                             |

>     |     min_val = min(encoded[])                             |

Reviewer 4: ```suggestion
    |     min_val = min(encoded[:])                             |
```
small nit show that it is meant to select all values, don't feel strongly.

Author: Applied in both diagram and table

    |     delta[i] = encoded[i] - min_val                      |
    +----------------------------------------------------------+
                              |
                              v
    +----------------------------------------------------------+
    |  4. BIT PACKING                                          |
    |     bit_width = ceil(log2(max_delta + 1))                |

>     |     Pack each delta into bit_width bits                  |

Reviewer 4: This is the same bitpacking scheme used in RLE/Bitpacked, we should make sure to make note of that.

Author: This is called out in the PackedValues section.

       +----------------+---------------+--------------+----+----+----+----+
       | compression    | integer       | log_vector   |     num_elements  |
       | _mode          | _encoding     | _size        |     (int32 LE)    |
       +----------------+---------------+--------------+----+----+----+----+
```

| Offset | Field | Size | Type | Description |
|--------|-------|------|------|-------------|
| 0 | compression_mode | 1 byte | uint8 | Compression mode (0 = ALP). Reserved for future variants (e.g., ALP-RD). |
| 1 | integer_encoding | 1 byte | uint8 | Integer encoding (must be 0 = FOR + bit-packing) |
| 2 | log_vector_size | 1 byte | uint8 | log2(vector\_size). Must be in the inclusive range \[3, 15\]. Recommended default: 10 (vector size 1024) |
| 3 | num_elements | 4 bytes | int32 | Total number of floating-point values in the page |

The number of vectors is `ceil(num_elements / vector_size)`. The last vector may
contain fewer than `vector_size` elements.

**Note:** The number of elements per vector is NOT stored in the header — it is
derived: `vector_size` for all vectors except the last, which may be smaller.

##### Offset Array

> Immediately following the header is an array of `num_vectors` little-endian uint32

Reviewer 4: Just want to make sure uint32 works here for java implementation (and we've tested it).  Would need to do some match but is possible to exceed int32 here based on other value restrictions?

Author: uint32 is safe — an offset can never exceed int32 range in a valid file. Each offset is a byte position within a single data page, and a Parquet page is capped at 2^31−1 bytes because PageHeader.uncompressed_page_size and compressed_page_size are both i32. So the high bit is never set, and reading offsets as signed int is always correct.

> values. Each offset gives the byte position of the corresponding vector's data,

Reviewer 4: Maybe include the formula.  `Vector Data Start = data page start (after thrift header) + 7 + offset`. Is this the right math?

Author: Added page_data_start + 7 + offset to Offset Array

```
 Byte:    0    1    2    3       4
       +----+----+----+----+-----------+
       | frame_of_reference | bit_width |
       |    (int32 LE)      |  (uint8)  |
       +----+----+----+----+-----------+
```

| Offset | Field | Size | Type | Description |
|--------|-------|------|------|-------------|
| 0 | frame_of_reference | 4 bytes | int32 | Minimum encoded integer in the vector |
| 4 | bit_width | 1 byte | uint8 | Bits per packed value. Range: \[0, 32\]. |

###### ForInfo for DOUBLE (9 bytes)

```
 Byte:    0    1    2    3    4    5    6    7       8
       +----+----+----+----+----+----+----+----+-----------+
       |          frame_of_reference           | bit_width |
       |              (int64 LE)               |  (uint8)  |
       +----+----+----+----+----+----+----+----+-----------+
```

| Offset | Field | Size | Type | Description |
|--------|-------|------|------|-------------|
| 0 | frame_of_reference | 8 bytes | int64 | Minimum encoded long in the vector |
| 8 | bit_width | 1 byte | uint8 | Bits per packed value. Range: \[0, 64\]. |

###### PackedValues

The FOR-encoded deltas, bit-packed into `ceil(num_elements_in_vector * bit_width / 8)` bytes.

> Values are bit-packed using the same LSB-first packing order as the

Reviewer 4: I think this addresses my concern above.  Not sure if it is worth mentioned padding the last byte with 0s

Author: Noted final byte zero-padded in MSBs

+-------------------------------------------------------------------+
```

The encoding uses two separate multiplications (not a single multiplication by
`10^(e-f)`, and not division) to ensure that implementations produce identical
floating-point results. All implementations MUST use the exact same floating-point
arithmetic and power-of-10 constants to guarantee cross-language interoperability.
The power-of-10 constants MUST be the correctly-rounded IEEE 754 values of the
decimal literals `1e0`, `1e1`, ..., `1e18` and `1e-1`, `1e-2`, ..., `1e-18` as
defined by the decimal-to-binary conversion in IEEE 754-2008 §5.12.2.
Implementations MUST NOT compute these constants at runtime via `pow()` or
equivalent functions, which are not guaranteed to be correctly rounded.

##### Fast Rounding

The `fast_round` function uses a "magic number" technique for branchless rounding.

`fast_round(value)` is defined as follows:

| Type   | Magic Number                      | Formula (value &ge; 0)           | Formula (value &lt; 0)           |
|--------|-----------------------------------|----------------------------------|----------------------------------|

> | FLOAT  | 2^22 + 2^23 = 12,582,912         | `(int32_t)((value + magic) - magic)` | `(int32_t)((value - magic) + magic)` |

Reviewer 4: maybe just note the outline above that the exact value is covered in a section below to address the branching.

Author: Done

| DOUBLE | 2^51 + 2^52 = 6,755,399,441,055,744 | `(int64_t)((value + magic) - magic)` | `(int64_t)((value - magic) + magic)` |

The sign branching is necessary because the technique relies on `value ± magic`

> landing in a binade where the unit in the last place (ULP) equals 1.0. For

Reviewer 4: ```suggestion
landing in a [binade](https://en.wikipedia.org/wiki/Binade) where the unit in the last place (ULP) equals 1.0. For
```
Had to look this up.

Author: Added Wikipedia link

|-------|---------|------------------------|---------|------------------------------------|------------|
| 0     | 1500.0  | 15000.0                | 15000   | 1500.0                             | No         |
| 1     | NaN     | -                      | -       | -                                  | Yes (NaN)  |
| 2     | 2500.0  | 25000.0                | 25000   | 2500.0                             | No         |
| 3     | 333.3   | 3333.0                 | 3333    | 333.3                              | No         |

**Step 2: Handle Exceptions**

Exception positions: \[1\]
Exception values: \[NaN\]
Placeholder: 15000 (first non-exception encoded value)
Encoded with placeholders: \[15000, 15000, 25000, 3333\]

**Step 3: Frame of Reference**

| Encoded            | min = 3333 | Delta |
|--------------------|------------|-------|
| 15000              | -          | 11667 |
| 15000 (placeholder)| -          | 11667 |
| 25000              | -          | 21667 |
| 3333               | -          | 0     |

**Step 4: Bit Packing**

max\_delta = 21667, bit\_width = ceil(log2(21668)) = 15 bits,

> packed\_size = ceil(4 * 15 / 8) = 8 bytes

Reviewer 4: should we include the hex-values here to avoid ambiguity?

Author: I'd lean against a hex table here. The spec already pins the constants unambiguously: they must be the correctly-rounded IEEE 754 conversions of the decimal literals 1e0..1e18 and 1e-1..1e-18 per §5.12.2, and MUST NOT be computed via pow(). Those are short literals, so every conforming decimal-to-binary conversion yields identical bit patterns — a hex table would be redundant and is itself an error-prone thing to maintain across 37 constants. If reviewers still want the exact patterns spelled out, I'd suggest an appendix/table in the eventual standalone ALP file rather than inline here. Does that work, or would you prefer the hex table now?

Reviewer: I agree adding hex values doesn't seem like it would add much

```
Bytes  AA 00 A3 BB 11 B4 CC 22 C5 DD 33 D6
```

<a name="ALP"></a>

> ### Adaptive Lossless floating-Point: (ALP = 10)

Reviewer 3: Just surprised that there isn't any change in the `parquet.thrift`?

Author: Done, fixed. I think I lost it during a recent rebase.

Reviewer: addressed in https://github.com/apache/parquet-format/pull/557/commits/47bde32ba89c9793fc89b88b691f54874cad0309

Supported Types: FLOAT, DOUBLE

This encoding is adapted from the paper
["ALP: Adaptive Lossless floating-Point Compression"](https://dl.acm.org/doi/10.1145/3626717)

> by Afroozeh and Boncz (SIGMOD 2024).

Reviewer 3: It seems that the paper has THREE authors but we have only listed TWO.

ALP works by converting floating-point values to integers using decimal scaling
(controlled by an *exponent* `e` and *factor* `f`), then applying Frame of
Reference (FOR) encoding and bit-packing. Values that cannot be losslessly
converted are stored separately as *exceptions*. The encoding achieves high
compression for decimal-like floating-point data (e.g., monetary values, sensor
readings) while remaining fully lossless. Each value is encoded independently,

> enabling random access to individual values and parallel encode/decode.

Reviewer 3: ```suggestion
enabling random access to individual values and parallel encoding/decoding.
```

#### Overview

ALP encoding consists of a page-level header followed by an offset array and one
or more encoded vectors (batches of values). Each vector contains up to
`vector_size` elements (default 1024).

```
+-------------+-----------------------------+--------------------------------------+
|   Header    |        Offset Array         |            Vector Data               |
|  (7 bytes)  |   (num_vectors * 4 bytes)   |            (variable)                |
+-------------+------+------+-----+---------+----------+----------+-----+----------+
| Page Header | off0 | off1 | ... | off N-1 | Vector 0 | Vector 1 | ... | Vec N-1  |
|  (7 bytes)  | (4B) | (4B) |     |  (4B)   |(variable)|(variable)|     |(variable)|
+-------------+------+------+-----+---------+----------+----------+-----+----------+
```

> The compression pipeline below describes *one* way to produce a conforming

Reviewer 3: IIUC, the informative compression description should be move to a separate doc later in the followup PR because here we just need necessary description for decoding a ALP-encoded page (i.e. the Page Layout section).

cc @alamb

Author: Ack. Makes sense. We can do this in a follow up PR

Reviewer: tracking in 
- https://github.com/apache/parquet-format/issues/602

vector. It is informative, not normative: an encoder may use any strategy as long
as it emits the byte layout defined in [Page Layout](#page-layout). Only that
byte layout and the [Decoding](#decoding) procedure are normative.

```
                    Input: float/double array
                              |
                              v
    +----------------------------------------------------------+
    |  1. CHOOSE PARAMETERS                                    |

>     |     Select (exponent, factor) pair for this vector       |

Reviewer 3: ```suggestion
    |     Select (exponent, factor) pair for this array        |
```

Be consistent with `Input: float/double array` above.

    +----------------------------------------------------------+
                              |
                              v
    +----------------------------------------------------------+
    |  2. DECIMAL ENCODING                                     |
    |     encoded[i] = fast_round(value[i] * 10^e * 10^(-f))  |
    |     Detect exceptions where decode(encode(v)) != v       |
    +----------------------------------------------------------+
                              |
                              v
    +----------------------------------------------------------+
    |  3. FRAME OF REFERENCE (FOR)                             |
    |     min_val = min(encoded[:])                            |

>     |     delta[i] = encoded[i] - min_val                      |

Reviewer 3: We need to clarify that `delta[i]` is an unsigned integer to avoid signed overflow.

Author: Thanks. Updated.

    +----------------------------------------------------------+
                              |
                              v
    +----------------------------------------------------------+
    |  4. BIT PACKING                                          |
    |     bit_width = ceil(log2(max_delta + 1))                |
    |     Pack each delta into bit_width bits                  |
    +----------------------------------------------------------+
                              |
                              v
                   Output: Serialized vector bytes
```

The `fast_round` used in step 2 is defined precisely in
[Fast Rounding](#fast-rounding) below, including the exact magic-number values
and the sign branching required for negative values.

#### Page Layout

##### Header (7 bytes)

All multi-byte values are stored in little-endian order.

```
 Byte:    0              1               2              3    4    5    6

>        +----------------+---------------+--------------+----+----+----+----+

Reviewer 3: ```suggestion
 Byte:         0                1              2         3    4    5    6
       +----------------+---------------+--------------+----+----+----+----+
```

Let's adjust the alignment here.

       | compression    | integer       | log_vector   |     num_elements  |
       | _mode          | _encoding     | _size        |     (int32 LE)    |
       +----------------+---------------+--------------+----+----+----+----+
```

| Offset | Field | Size | Type | Description |
|--------|-------|------|------|-------------|
| 0 | compression_mode | 1 byte | uint8 | Compression mode (0 = ALP). Reserved for future variants (e.g., ALP-RD). |
| 1 | integer_encoding | 1 byte | uint8 | Integer encoding (must be 0 = FOR + bit-packing) |

> | 2 | log_vector_size | 1 byte | uint8 | log2(vector\_size). Must be in the inclusive range \[3, 15\]. Recommended default: 10 (vector size 1024) |

Reviewer 3: If the minimum value is 3, why not defining it as `log_vector_size_minus_three` so it can start with 0? Not a strong opinion since this anyway consume a byte which is large enough.

Author: I'd like to keep log_vector_size, maps directly to 1 << log_vector_size.

> | 3 | num_elements | 4 bytes | int32 | Total number of floating-point values in the page |

Reviewer 3: ```suggestion
| 3 | num_elements | 4 bytes | int32 | Total number of non-null floating-point values in the page |
```

The number of vectors is `ceil(num_elements / vector_size)`. The last vector may
contain fewer than `vector_size` elements.

**Note:** The number of elements per vector is NOT stored in the header — it is
derived: `vector_size` for all vectors except the last, which may be smaller.

##### Offset Array

Immediately following the header is an array of `num_vectors` little-endian uint32
values. Each offset gives the byte position of the corresponding vector's data,
measured from the start of the offset array itself.

> The first offset always equals `num_vectors * 4` (pointing just past the offset array).

Reviewer 3: Why `num_vectors * 4` but not `num_vectors * 4 + 7` to include the header? Usually the whole buffer is stored including header, so implementations need to manually add 7 each time to get an absolute offset to each vector.

Reviewer: I don't see any particular reason for one or the other. Given that we have used the `num_vectors * 4` in the prototype implementation I think it is ok to leave this as is

Each subsequent offset equals the previous offset plus the stored size of the
previous vector. No padding is inserted between vectors.

> A vector's absolute byte position within the page is

Reviewer 3: This is incorrect when pages are compressed or have rep/def levels. I would suggest removing this sentence. If we want to emphasize random access of ALP, perhaps we should replace `page_data_start` with a more precise start point after decompressed.

`page_data_start + 7 + offset`, where `page_data_start` is the first byte after
the page's Thrift header and `7` is the size of the ALP header.

##### Vector Format

> Each vector is self-describing and contains the encoding parameters, FOR metadata,

Reviewer 3: This only applies to `compression_mode=0` and `integer_encoding=0`, right? Should we be explicit on this? Future settings may not even have `AlpInfo` and `ForInfo`.

Author: Done. Made it more explicit.

bit-packed encoded values, and exception data.

```

> +-------------------+-----------------+-------------------+---------------------+-------------------+

Reviewer 3: Can we enhance the graph by pointing out areas of vector header and data section?

Author: Updated.

> |      AlpInfo      |     ForInfo     |   PackedValues    | ExceptionPositions  | ExceptionValues   |

Reviewer 3: Why here we use camel case but other places use snake case?

Author: In thrift I we have composite structures as camel and scalars as snake.

|     (4 bytes)     | (5B or 9B)      |    (variable)     |     (variable)      |    (variable)     |
+-------------------+-----------------+-------------------+---------------------+-------------------+
```

Vector header sizes:
| Type   | AlpInfo | ForInfo | Total Header |
|--------|---------|---------|--------------|
| FLOAT  | 4 bytes | 5 bytes | 9 bytes      |
| DOUBLE | 4 bytes | 9 bytes | 13 bytes     |

Data section sizes:
| Section             | Size Formula                | Description                  |
|---------------------|-----------------------------|------------------------------|
| PackedValues        | ceil(num\_elements\_in\_vector * bit\_width / 8) | Bit-packed delta values      |

> | ExceptionPositions  | num\_exceptions * 2 bytes   | uint16 indices of exceptions |

Reviewer 3: `num_exceptions` is described below so readers may have no idea where it comes from when reading this line.

Author: Added forward reference.

> | ExceptionValues     | num\_exceptions * sizeof(encoded type) (float=4 and double=8) | Original float/double values |

Reviewer 3: In the past, parquet-java has NaN canonicalization which is wrong. Should we clarify that these are raw bits?

Author: The spec requires exception values be the exact IEEE-754 bit pattern and must not canonicalize NaN. Added a cross-reference.

|                                                                   |
|   encoded = fast_round( value  *  10^e  *  10^(-f) )             |
|                                                                   |
|   decoded = encoded  *  10^f  *  10^(-e)                          |
|                                                                   |
+-------------------------------------------------------------------+
```

The encoding uses two separate multiplications (not a single multiplication by
`10^(e-f)`, and not division) to ensure that implementations produce identical
floating-point results. All implementations MUST use the exact same floating-point
arithmetic and power-of-10 constants to guarantee cross-language interoperability.
The power-of-10 constants MUST be the correctly-rounded IEEE 754 values of the
decimal literals `1e0`, `1e1`, ..., `1e18` and `1e-1`, `1e-2`, ..., `1e-18` as
defined by the decimal-to-binary conversion in IEEE 754-2008 §5.12.2.
Implementations MUST NOT compute these constants at runtime via `pow()` or
equivalent functions, which are not guaranteed to be correctly rounded.

##### Fast Rounding

> The `fast_round` function uses a "magic number" technique for branchless rounding.

Reviewer 5: It's not branchless if there's a sign branching involved.

We should also explain why this function is useful or necessary. What is the expected rounding mode? Halfway to even?

Reviewer: > It's not branchless if there's a sign branching involved.

Yes, this is confusing. The original paper (and the last time I reviewed this spec implementation) did not have this branching. I don't the branching is necessary and the explanation is very confusing (if not  incorrect -- see my comments on this thread https://github.com/apache/parquet-format/pull/557#discussion_r3292628978)

> We should also explain why this function is useful or necessary. What is the expected rounding mode? Halfway to even?

I believe the rationale is to "recover the intended integer after computing `value * 10^e * 10^-f` which introduces mathematical noise due to floating point arithmetic

This is not well explained in the original paper either, which perhaps lead to the confusion in this section

From my reading,  this part of the spec is normative anyway. Since it isn't called on the decode path,  encoders can use a different method (like the original branchless  version) and still produce conformant files

Thus I don't think my concerns with this section are blocking; However, as written I now find it quite misleading and it would be good to clarify / fix this

`fast_round(value)` is defined as follows:

| Type   | Magic Number                      | Formula (value &ge; 0)           | Formula (value &lt; 0)           |
|--------|-----------------------------------|----------------------------------|----------------------------------|
| FLOAT  | 2^22 + 2^23 = 12,582,912         | `(int32_t)((value + magic) - magic)` | `(int32_t)((value - magic) + magic)` |
| DOUBLE | 2^51 + 2^52 = 6,755,399,441,055,744 | `(int64_t)((value + magic) - magic)` | `(int64_t)((value - magic) + magic)` |

> The sign branching is necessary because the technique relies on `value ± magic`

Reviewer: This doesn't make any sense to me. It seems like an overly complicated argument with fancy terms that ultimately isn't useful in the context of ALP and has no empirical justification

See my comment/question on the original thread here:  https://github.com/apache/parquet-format/pull/557#discussion_r3292628978

Reviewer 7: Here's an example where the branch-less fast round produces an error:

```java
@Test
  public void testFloats() {
    float f = 6194305.5f;
    float magic = 8388608.0f;

    assertThat((int) (f + magic - magic)).isEqualTo(6194306);

    float nf = -6194305.5f;
    float nmagic = -8388608.0f;

    assertThat((int) (nf + nmagic - nmagic)).isEqualTo(-6194306);

    float umagic = 12582912.0f;

    // off by 1
    assertThat((int) (nf + umagic - umagic)).isEqualTo(-6194305);
  }
```

I think the issue is limited to non-integer values in the bottom half of (-2^23, -2^22]. When using the mid-point constant to avoid branching, they are not forced into the 2^23 binade. As a result, the precision is not limited and the rounding does not occur.

Reviewer: > Here's an example where the branch-less fast round produces an error:

When I played around with it, I found that switching to branchless fast rounding simply trades one set of error cases for another 

I made a [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024) to show how to test various combinations work (repo here [https://github.com/alamb/fast_round](https://github.com/alamb/fast_round))

For example here are three cases where the branchless rounding is actually *better* and the signed version is worse
```rust
        // (b) At and above 2^52 plain is better
        Case {
            input: -5226231736140883.0,
            correct: -5226231736140883,
            fast_round: -5226231736140883,  // exact
            sign_branch: -5226231736140884, // off by 1
            description: "[2^52,2^53): plain is exact, branch corrupts it",
        },
        Case {
            input: -12298387206974542.0,
            correct: -12298387206974542,
            fast_round: -12298387206974542,  // exact
            sign_branch: -12298387206974544, // off by 2
            description: "[2^53,2^54): plain is exact, branch corrupts it",
        },
        Case {
            input: -143122344316276784.0,
            correct: -143122344316276784,
            fast_round: -143122344316276784,  // exact
            sign_branch: -143122344316276800, // off by 16
            description: "large: plain is exact, branch corrupts it",
        },
```

Reviewer: Update here is that @rdblue and I discussed this a bit more and the conclusion I think is that since this is normative (writers can make different choices) the current spec is fine as is. 

It is not clear to me if the different specific values at which the different fast_round implementations differ (and thus will incorrectly round) will be very common to me. I think we can see how the roll out goes and then go from there

Reviewer 4: If so, we should call this out in the spec, as above there is specific language on required parts of the arithmetic.

Author: I over indexed on Ryan's example and updated the wording, thanks Andrew for the correction.
Given the examples that Andrew has brought up i'll keep the wording but instead of "encoder must do it", I can change it to "one of the ways of doing it is..."

```
Bytes  AA 00 A3 BB 11 B4 CC 22 C5 DD 33 D6
```

<a name="ALP"></a>
### Adaptive Lossless floating-Point: (ALP = 10)

Supported Types: FLOAT, DOUBLE

This encoding is adapted from the paper
["ALP: Adaptive Lossless floating-Point Compression"](https://dl.acm.org/doi/10.1145/3626717)

> by Afroozeh, Kuffo, and Boncz (SIGMOD 2024).

Author: Fixed

is normative: to reconstruct a value every reader MUST compute
`decoded = encoded * 10^f * 10^(-e)` using the same two-step multiplication and the
same power-of-10 constants, so that all implementations reproduce the stored value
bit-for-bit. The power-of-10 constants MUST be the correctly-rounded IEEE 754
values of the decimal literals `1e0`, `1e1`, ..., `1e18` and `1e-1`, `1e-2`, ...,
`1e-18` as defined by the decimal-to-binary conversion in IEEE 754-2008 §5.12.2.
Implementations MUST NOT compute these constants at runtime via `pow()` or
equivalent functions, which are not guaranteed to be correctly rounded.

The **encode** direction — mapping each value to the integer it will be stored as,
via `fast_round(value * 10^e * 10^(-f))` — is informative, not normative. An
encoder MAY choose that integer by any means, because every value is checked
against the normative decode above and any value that does not round-trip exactly
is stored as an exception. The rounding method therefore affects only compression
ratio and exception count, never correctness or what a reader decodes. The
`fast_round` technique below is one recommended implementation.

> ##### Fast Rounding (informative)

Author: @alamb do you recommend I simplify the language here or remove part of the section?

Author: I'm going for simplification for now

directly in a single branch-free form; others add a sign test so that both positive
and negative values land in a binade with ULP 1.0:

| Type   | Magic Number                      | Formula (value &ge; 0)           | Formula (value &lt; 0)           |
|--------|-----------------------------------|----------------------------------|----------------------------------|
| FLOAT  | 2^23 = 8,388,608                 | `(int32_t)((value + magic) - magic)` | `(int32_t)((value - magic) + magic)` |
| DOUBLE | 2^52 = 4,503,599,627,370,496     | `(int64_t)((value + magic) - magic)` | `(int64_t)((value - magic) + magic)` |

The `value ± magic` operations are floating-point, not integer, arithmetic; only
the final cast converts to an integer. The arithmetic should be evaluated in the
value's own precision (FLOAT in binary32, DOUBLE in binary64), because it relies on
`value ± magic` landing in a binade where the ULP equals 1.0, which holds only in
the matching precision — evaluating a FLOAT computation in double precision, for
example, places `8,388,608` in a binade whose ULP is far below 1.0 and no rounding
occurs.

> Each form has a limited valid domain — roughly (-2^23, 2^23) for floats and

Reviewer: I like this wording
